libc-bottom-half/crt/crt1-command.c

```diff
@@ -3,10 +3,63 @@
 extern void __wasi_init_tp(void);
 #endif
 #include <wasi/api.h>
+#include <stdlib.h>
+#include <sysexits.h>
 extern void __wasm_call_ctors(void);
 extern int __main_void(void);
 extern void __wasm_call_dtors(void);
 
+// The user's `main` function, expecting arguments.
+int __main_argc_argv(int argc, char *argv[]);
+
+// If the user's `main` function expects arguments, the compiler will rename
+// it to `__main_argc_argv`, and this version will get linked in, which
+// initializes the argument data and calls `__main_argc_argv`.
+__attribute__((__weak__, nodebug))
+int __main_void(void) {
+    __wasi_errno_t err;
+
+    // Get the sizes of the arrays we'll have to create to copy in the args.
+    size_t argv_buf_size;
+    size_t argc;
+    err = __wasi_args_sizes_get(&argc, &argv_buf_size);
+    if (err != __WASI_ERRNO_SUCCESS) {
+        _Exit(EX_OSERR);
+    }
+
+    // Add 1 for the NULL pointer to mark the end, and check for overflow.
+    size_t num_ptrs = argc + 1;
+    if (num_ptrs == 0) {
+        _Exit(EX_SOFTWARE);
+    }
+
+    // Allocate memory for storing the argument chars.
+    char *argv_buf = malloc(argv_buf_size);
+    if (argv_buf == NULL) {
+        _Exit(EX_SOFTWARE);
+    }
+
+    // Allocate memory for the array of pointers. This uses `calloc` both to
+    // handle overflow and to initialize the NULL pointer at the end.
+    char **argv = calloc(num_ptrs, sizeof(char *));
+    if (argv == NULL) {
+        free(argv_buf);
+        _Exit(EX_SOFTWARE);
+    }
+
+    // Fill the argument chars, and the argv array with pointers into those chars.
+    // TODO: Remove the casts on `argv_ptrs` and `argv_buf` once the witx is updated with char8 support.
+    err = __wasi_args_get((uint8_t **)argv, (uint8_t *)argv_buf);
+    if (err != __WASI_ERRNO_SUCCESS) {
+        free(argv_buf);
+        free(argv);
```

Comment:
Unrelated to this change but I wonder why we don't use `alloca` here to avoid pulling in malloc?  

The emscripten version does this: https://github.com/emscripten-core/emscripten/blob/5ab750c822e47b0878e939f32d0f297edf632cf0/system/lib/standalone/__main_void.c#L41

Reply:
iirc the default stack size set by `wasm-ld` is 65536 which is exceptionally small, so using that for `argv_buf` sounds pretty risky.

Reply:
I don't think the fact that we have a small default stack size should be used as an argument for not using the stack in places where is makes sense (like this one).

Folks who want to accept long argv lists can easily link with `-Wl,-zstack-size=XX`.    I think the downsize of linking in a malloc implementation into all programs high enough to warrant making this change.

Reply:
Are you suggesting some kind of auto-detection of the need to real malloc vs a bump allocator?   Or some kind of opt-in for folks who think they only need a bump allocator?  (e.g. `-lfake-malloc`?)

Reply:
I guess we could do it with weak symbols.

Code like this could call `maybe_malloc` and `maybe_free`, which are provides weakly by `bumpmalloc.c` but strongly as aliases of `malloc/free` in `dlmalloc.c`?   That might work... but would require those aliases to be added to all malloc/free providers (dlmalloc being one of many)

Reply:
I was thinking more like just `-lfake-malloc`.

Reply:
I think that would be unfortunate since it would result is a "bloated-by-default" outcome for small programs.

Reply:
On the other hand many such small programs (.e.g hello world) are not that important since they don't represent real world uses?

Reply:
I expect most real-world programs call `malloc` somewhere anyway. The other side of it is, if we had more people maintaining wasi-libc, more things would become possible.